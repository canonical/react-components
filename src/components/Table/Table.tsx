import classNames from "classnames";
import React, { HTMLProps, ReactNode } from "react";

import type { ClassName, PropsWithSpread } from "types";

export type Props = PropsWithSpread<
  {
    /**
     * The content of the table.
     */
    children?: ReactNode;
    /**
     * Optional class(es) to pass to the wrapping table element.
     */
    className?: ClassName;
    /**
     * Whether the table can expand hidden cells.
     */
    expanding?: boolean;
    /**
     * Whether the table should show card styling on smaller screens.
     */
    responsive?: boolean;
  },
  HTMLProps<HTMLTableElement>
>;

const Table = ({
  children,
  className,
  expanding = false,
  responsive = false,
  ...props
}: Props): JSX.Element => (
  <table
    role="grid"
    className={classNames(className, {
      "p-table--mobile-card": responsive,
      "p-table--expanding": expanding,
    })}
    {...props}
  >
    {children}
  </table>
);

export default Table;
