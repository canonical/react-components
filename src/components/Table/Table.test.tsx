import { render, screen } from "@testing-library/react";
import React from "react";

import Table from "./Table";

const TEST_ID = "table";

describe("Table", () => {
  it("renders", () => {
    render(
      <Table data-testid={TEST_ID}>
        <thead></thead>
      </Table>
    );
    expect(screen.getByTestId(TEST_ID)).toMatchSnapshot();
  });

  it("can be expanding", () => {
    render(<Table data-testid={TEST_ID} expanding={true} />);
    expect(screen.getByTestId(TEST_ID)).toHaveClass("p-table--expanding");
  });

  it("can be responsive", () => {
    render(<Table data-testid={TEST_ID} responsive={true} />);
    expect(screen.getByTestId(TEST_ID)).toHaveClass("p-table--mobile-card");
  });

  it("can pass additional classes", () => {
    render(
      <Table data-testid={TEST_ID} responsive={true} className="extra-class" />
    );
    expect(screen.getByTestId(TEST_ID)).toHaveClass("p-table--mobile-card");
    expect(screen.getByTestId(TEST_ID)).toHaveClass("extra-class");
  });
});
