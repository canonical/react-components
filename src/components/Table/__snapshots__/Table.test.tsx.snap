// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders 1`] = `
<table
  className=""
  role="grid"
>
  <thead />
</table>
`;
